File: Python/phate.py
#!/usr/bin/env python3

######
#
# Code translated from Matlab to Python by Daniel Burkhardt. Original code by Kevin Moon, Yale Univeristy.
# Author:    Daniel B. Burkhardt, Yale University
# Created:   03.24.2017
# 
######


import numpy as np
import sklearn
from sklearn.neighbors import NearestNeighbors
from sklearn.manifold import MDS
from scipy.spatial.distance import pdist
from scipy.spatial.distance import squareform
import matplotlib.pyplot as plt

from __future__ import division

def cmdscale(D):
    """                                                                                       
    Classical multidimensional scaling (MDS) by Francis Song, New York University
    http://www.nervouscomputer.com/hfs/cmdscale-in-python/                                                  
                                                                                               
    Parameters                                                                                
    ----------                                                                                
    D : (n, n) array                                                                          
        Symmetric distance matrix.                                                            
                                                                                               
    Returns                                                                                   
    -------                                                                                   
    Y : (n, p) array                                                                          
        Configuration matrix. Each column represents a dimension. Only the                    
        p dimensions corresponding to positive eigenvalues of B are returned.                 
        Note that each dimension is only determined up to an overall sign,                    
        corresponding to a reflection.                                                        
                                                                                               
    e : (n,) array                                                                            
        Eigenvalues of B.                                                                     
                                                                                               
    """
    # Number of points                                                                        
    n = len(D)
 
    # Centering matrix                                                                        
    H = np.eye(n) - np.ones((n, n))/n
 
    # YY^T                                                                                    
    B = -H.dot(D**2).dot(H)/2
 
    # Diagonalize                                                                             
    evals, evecs = np.linalg.eigh(B)
 
    # Sort by eigenvalue in descending order                                                  
    idx   = np.argsort(evals)[::-1]
    evals = evals[idx]
    evecs = evecs[:,idx]
 
    # Compute the coordinates using positive-eigenvalued components only                      
    w, = np.where(evals > 0)
    L  = np.diag(np.sqrt(evals[w]))
    V  = evecs[:,w]
    Y  = V.dot(L)
 
    return Y, evals


### Phate code with toy example. The data matrix `M` should have `N` rows and `d` columns. 
### If `d` is greater than 100, then PCA is reccomended as a dimensionality reduction strategy.

### Generating random fractal tree via DLA
n_dim = 100
n_branch = 20
branch_length = 100
n_drop = 0
rand_multiplier = 2
np.random.seed(37)
sigma = 4

### Loading / Generating toy data
if M is None:
	try:
	    M = np.genfromtxt('test_M.csv',delimiter = ',')
	except OSError:
	    M = np.cumsum(-1 + rand_multiplier*np.random.rand(branch_length,n_dim),0)
	    for i in range(n_branch-1):
	        ind = np.random.randint(branch_length)
	        new_branch = np.cumsum(-1 + rand_multiplier*np.random.rand(branch_length,n_dim),0)
	        M = np.concatenate([M,new_branch+M[ind,:]])
	    noise = np.random.normal(0, sigma,M.shape)
	    M = M + noise
	    np.savetxt('test_M.csv',M,delimiter=',')

C = [i//n_branch for i in range(n_branch*branch_length)] #returns the group labels for each point to make it easier to visualize embeddings

### Application of diffusion maps

a = 10 #determines the decay rate of the kernel tails
k = 5 #this is used for autotuning the kernel bandwidth
t = 30 #how much diffusion to perform

nbrs = NearestNeighbors(n_neighbors=k+1).fit(M)
knnDST, indices = nbrs.kneighbors(M)
epsilon = knnDST[:,k] # bandwidth(x) = distance to k-th neighbor of x
PDX = squareform(pdist(M))
PDX = (PDX / epsilon).T # autotuning d(x,:) using epsilon(x). not sure why the .T is needed yet
GsKer = np.exp(-1 * ( PDX ** a)) # not really Gaussian kernel
GsKer = GsKer + GsKer.T
DiffDeg = np.diag(np.sum(GsKer,0)) # degrees
DiffOp = np.dot(np.diag(np.diag(DiffDeg)**(-1)),GsKer) # row stochastic
DiffAff = np.dot(np.dot(np.diag(np.diag(DiffDeg)**(-1/2)),GsKer),np.diag(np.diag(DiffDeg)**(-1/2))); # symmetric conjugate affinities
DiffAff = (DiffAff + DiffAff.T)/2; #clean up numerical inaccuracies to maintain symmetry

Phi, Lambda, V = np.linalg.svd(DiffAff) #eigenvalues & eigenvectors of affinities
Lambda = np.diag(Lambda)
Phi = (Phi.T / Phi[:,0]).T  #conjugate back to eigenvectors of row stochastic
Diffmap = lambda t: np.dot(Phi[:,1:len(Phi)],Lambda[1:len(Lambda),1:len(Lambda)]**t) #time-indexed diffusion maps

X = np.linalg.matrix_power(DiffOp,t)
X[X == 0] = np.finfo(float).eps
X = -1*np.log(X)

## MDS embeddings, each gives a different output.
X_dist = squareform(pdist(X))

#classical MDS as defined above
Y_cmds = cmdscale(X_dist)[0]
#Metric MDS from sklearn
Y_mmds = MDS(n_components=2, metric=True, max_iter=3000, eps=1e-12,
                    dissimilarity="precomputed", random_state=seed, n_jobs=16,
                    n_init=1).fit_transform(X_dist)
#Nonmetric MDS from sklearn using metric MDS as an initialization
nmds_init = Y_mmds
Y_nmds = MDS(n_components=2, metric=False, max_iter=3000, eps=1e-12,
                    dissimilarity="precomputed", random_state=seed, n_jobs=16,
                    n_init=1).fit_transform(X_dist,init=nmds_init)

##to visualize embedding if desired
Y = Y_cmds
plt.scatter(Y[:,0],Y[:,1],c=C,cmap='cubehelix')
y_min = min(Y[:,1]) * 1.1
y_max = max(Y[:,1]) * 1.1
plt.ylim(y_min,y_max)

File: Matlab/phate.m
function [Y, DiffOp, DiffOp_t, Aff] = phate(data, varargin)
% Runs PHATE on input data. data must have cells on the rows and genes on the columns
%
% Authors: Kevin Moon, David van Dijk
% Created: March 2017

% OUTPUT
%       Y = the PHATE embedding
%       DiffOp = the diffusion operator which can be used as optional input with another run
%       DiffOp_t = DiffOp^t

% INPUT
%       data = data matrix. Must have cells on the rows and genes on the columns
% varargin:
%   't' (default = 20)
%       Diffusion time scale
%   'k' (default = 5)
%       k for the adaptive kernel bandwidth
%   'a' (default = 10)
%       The alpha parameter in the exponent of the kernel function. Determines the kernel decay rate
%   'ndim' (default = 2)
%       The number of desired PHATE dimensions in the output Y. 2 or 3 is best for visualization. A higher number can be used for
%       running other analyses on the PHATE dimensions.
%   'pca_method' (default = 'random')
%       The desired method for implementing pca for preprocessing the data. Options include 'svd', 'random', and 'none' (no pca)
%   'npca' (default = 100)
%       The number of PCA components for preprocessing the data
%   'mds_method' (default = 'cmds_fast')
%       Method for implementing MDS. Choices are 'cmds' (built-in matlab function), 'cmds_fast' (uses fast PCA), 'nmmds' (nonmetric MDS),
%       and 'mmds' (metric MDS)
%   'distfun' (default = 'euclidean')
%       The desired distance function for calculating pairwise distances on the data.
%   'distfun_mds' (default = 'euclidean')
%       The desired distance function for MDS. Choices are 'euclidean',
%       'cosine'.
%   'DiffOp' (default = [])
%       If the diffusion operator has been computed on a prior run with the desired parameters, then this option can be used to
%       directly input the diffusion operator to save on computational time.
%   'DiffOp_t' (default = [])
%       Same as for 'DiffOp', if the powered diffusion operator has been computed on a prior run with the desired parameters,
%       then this option can be used to directly input the diffusion operator to save on computational time.
%   'Aff' (default = [])
%       Kernel / affinity matrix, e.g. to embed a graph. The default ([]) is to compute one from
%       the supplied data.

% set up default parameters
k = 5;
a = 10;
npca = 100;
t = 20;
ndim = 2;
mds_method = 'cmds_fast';
distfun = 'euclidean';
distfun_mds = 'euclidean';
pca_method = 'random';
DiffOp = [];
DiffOp_t = [];
Aff = [];

% get input parameters
for i=1:length(varargin)
    % adaptive k-nn bandwidth
    if(strcmp(varargin{i},'k'))
       k =  lower(varargin{i+1});
    end
    % alpha parameter for kernel decay rate
    if(strcmp(varargin{i},'a'))
       a =  lower(varargin{i+1});
    end
    % diffusion time
    if(strcmp(varargin{i},'t'))
       t =  lower(varargin{i+1});
    end
    % Number of pca components
    if(strcmp(varargin{i},'npca'))
       npca =  lower(varargin{i+1});
    end
    % Number of dimensions for the PHATE embedding
    if(strcmp(varargin{i},'ndim'))
       ndim =  lower(varargin{i+1});
    end
    % Method for MDS
    if(strcmp(varargin{i},'mds_method'))
       mds_method =  lower(varargin{i+1});
    end
    % Distance function for the inputs
    if(strcmp(varargin{i},'distfun'))
       distfun =  lower(varargin{i+1});
    end
    % distfun for MDS
    if(strcmp(varargin{i},'distfun_mds'))
       distfun_mds =  lower(varargin{i+1});
    end
    % Method for PCA
    if(strcmp(varargin{i},'pca_method'))
       pca_method =  lower(varargin{i+1});
    end
    % Use precomputed diffusion operator?
    if(strcmp(varargin{i},'DiffOp'))
       DiffOp = lower(varargin{i+1});
    end
    % Use precomputed powered diffusion operator?
    if(strcmp(varargin{i},'DiffOp_t'))
       DiffOp_t = lower(varargin{i+1});
    end
    % Use precomputed kernel / affinity matrix
    if(strcmp(varargin{i},'Aff'))
       Aff = lower(varargin{i+1});
    end
end

disp '======= PHATE ======='

% Check to see if precomputed DiffOp or DiffOp_t are given

if(isempty(DiffOp) && isempty(DiffOp_t))
    % check if Aff is supplied
    if isempty(Aff)
        if ~strcmp(pca_method, 'none')
            M = svdpca(data, npca, pca_method);
        else
            M = data;
        end
        
        disp 'computing distances'
        if isa(distfun, 'function_handle')
            PDX = pdist2(M, M, distfun);
        else
            PDX = squareform(pdist(M, distfun));
        end
        
        disp 'computing kernel and operator'
        knnDST = sort(PDX);
        epsilon = knnDST(k+1,:); % bandwidth(x) = distance to k-th neighbor of x
        PDX = bsxfun(@rdivide,PDX,epsilon); % autotuning d(x,:) using epsilon(x)
        Aff = exp(-PDX.^a); % affinity matrix
        Aff = Aff + Aff'; % make sure it's symmetric
    end
    DiffDeg = diag(sum(Aff,2)); % degrees
    DiffOp = DiffDeg^(-1)*Aff; % row stochastic

    % Clear a bit of space for memory
    clear PDX DiffDeg
end

% Check to see if pre computed DiffOp_t is given
if(isempty(DiffOp_t))
    disp 'diffusing operator'
    DiffOp_t = DiffOp^t;
    
end
X = DiffOp_t;

disp 'potential recovery'
X(X<=eps)=eps;
X = -log(X);

disp(['MDS distfun: ' distfun_mds])
if ~strcmp(distfun_mds, 'none')
    if strcmp(distfun_mds, 'euclidean')
        if ~strcmp(pca_method, 'none')
            X = svdpca(X, npca, pca_method); % to make pdist faster
        end
    end
    X = squareform(pdist(X, distfun_mds));
end

switch mds_method
    % CMDS using fast pca
    case 'cmds_fast'
        disp 'Fast CMDS'
        Y = randmds(X, ndim);
    % built-in MATLAB version of CMDS
    case 'cmds'
        disp 'CMDS'
        Y = cmdscale(X, ndim);
    % built-in MATLAB version of NMMDS
    case 'nmmds'
        disp 'NMMDS'
        opt = statset('display', 'iter');
        Y_start = randmds(X, ndim);
        Y = mdscale(X, ndim, 'options', opt, 'start', Y_start);
    % built-in MATLAB version of metric MDS
    case 'mmds'
        disp 'MMDS'
        opt=statset('display', 'iter');
        Y_start=randmds(X,ndim);
        Y = mdscale(X,ndim,'options',opt,'Criterion','metricstress','start',Y_start);
end

disp 'done.'
